### ISSUE TITLE
Mouse Over

### ISSUE THREAD

**Issue by louisstow** [craftyjs/Crafty#67]
Mouse over only works when the mouse physically moves over the entity. It doesn't work when an entity moves over the mouse.

**Comment by CoderPuppy** [comment]
Maybe it could be fixed by keeping track of the entities and the mouse.

**Comment by louisstow** [comment]
The problem is it only looks for `MouseOver` events during `MouseMove`. Having to check on every frame would be very incredibly slow.

**Comment by atimmer** [comment]
Can't you see the mouse as an hidden entity which you then use collision detection to fire the events

**Comment by mpetrovic** [comment]
I think cheaper would be to store the mouse position, then whenever an entity with a mouseover event is moved on screen, check it against that position. 

The trick would be keeping track of all the ways the mouse would move relative to the page.

**Comment by louisstow** [comment]
The current method is much too slow. It has a mousemove handler and then
determines if it hit something (MouseOver). But the mousemove event is being
called many many times and is having to check the Spatial HashMap many
times. Would it be possible to trigger mousemove less often?

That is a good idea though, mpetrovic

**Comment by mpetrovic** [comment]
It's not that mousemove is triggering too much, it's that we're looping through every entity on every mousemove when we really don't need too. 

Why aren't we using DOM events and falling back to the map solution when the target is a canvas? And to speed up canvas, could we do something similar to redraw regions to limit the number of entities being scanned?

**Comment by louisstow** [comment]
No it uses the HashMap and looks for every entity in the same Hashed region
as the mouse position.

DOM events is what should be happening for DOM nodes.

**Comment by mpetrovic** [comment]
So canvas is already as fast as we can make it. We just need to get DOM events to be utilized properly.

**Comment by louisstow** [comment]
In theory it's faster. I do remember it slowing down when manically moving
the mouse over many Mouse items. Probably because it had to do about 50
searches in a span of a second.

Yes, definitely need to add DOM events.

**Comment by louisstow** [comment]
Could we store the last mouse position and determine if the new mouse
position is in the same Hashed region and therefore don't do anything... not
sure if it would break anything.

**Comment by mpetrovic** [comment]
I haven't looked at the Hash Map stuff, so I don't know what that would entail .We would be storing the mouse position anyway to handle the original issue, assuming it still applies when dom events are in play. We should probably store the last hash map entry itself rather then comparing 2 separate positions against the hash map.

**Comment by mpetrovic** [comment]
mouseDispatch gets the DOM entity target using properties on the event now. You'll have to handle optimizations for the canvas side.

**Comment by mpetrovic** [comment]
Just updated this a bit. If our target isn't an entity, roll up through the ancestor tree until we find one.

**Comment by ahilles107** [comment]
This bug was fixed or not?

**Comment by BossRighteous** [comment]
I'll take a look as I address the mobile touch events. The technique I implemented for MouseMove optimization was the caching of current position on MouseMove. MouseMoves are triggered per pixel so they occur far to frequently to do anything other than updating global position variables. A worthwhile solution is to incorporate an enterFrame style interval for the triggering of the MouseMove event to entities. That way it occurs per tick instead of per pixel. I saw significant gains in speed when using this approach elsewhere. I will dig deeper and report back if I find any optimizations.
